yarn-project/noir-contracts/src/contracts/token_contract/src/types.nr

```diff
@@ -0,0 +1,176 @@
+    }
+  }
+
+  fn serialize(self: Self) -> [Field; 1] {
+    [self.address]
+  }
+
+  fn deserialize(fields: [Field; 1]) -> Self {
+    Self {
+      address: fields[0]
+    }
+  }
+}
+
+struct EthereumAddress {
+  address: Field
+}
+
+impl EthereumAddress {
+  fn new(address: Field) -> Self {
+    Self {
+      address
+    }
+  }
+
+
+  fn serialize(self: Self) -> [Field; 1] {
+    [self.address]
+  }
+
+  fn deserialize(fields: [Field; 1]) -> Self {
+    Self {
+      address: fields[0]
+    }
+  }
+}
+
+
+
+// Transparent note represents a note that is created in the clear (public execution),
+// but can only be spent by those that know the preimage of the "secret_hash"
+struct TransparentNote {
+    amount: Field,
+    secret_hash: Field,
+    // the fields below are not serialised/deserialised
+    secret: Field,
+    header: NoteHeader,
+}
+
+impl TransparentNote {
+
+    // CONSTRUCTORS
+
+    fn new(amount: Field, secret_hash: Field) -> Self {
+        TransparentNote {
+            amount: amount,
+            secret_hash: secret_hash,
+            secret: 0,
+            header: NoteHeader::empty(),
+        }
+    }
+
+    // new oracle call primitive
+    // get me the secret corresponding to this hash
+    fn new_from_secret(amount: Field, secret: Field) -> Self {
+        TransparentNote {
+            amount: amount,
+            secret_hash: TransparentNote::compute_secret_hash(secret),
+            secret: secret,
+            header: NoteHeader::empty(),
+        }
+    }
+
+
+    // STANDARD NOTE_INTERFACE FUNCTIONS
+
+    fn serialise(self) -> [Field; TRANSPARENT_NOTE_LEN] {
+        [self.amount, self.secret_hash]
+    }
+
+    fn deserialise(preimage: [Field; TRANSPARENT_NOTE_LEN]) -> Self {
+        TransparentNote {
+            amount: preimage[0],
+            secret_hash: preimage[1],
+            secret: 0,
+            header: NoteHeader::empty(),
+        }
+    }
+
+    fn compute_note_hash(self) -> Field {
+        // TODO(#1205) Should use a non-zero generator index.
+        dep::std::hash::pedersen([
+            self.amount,
+            self.secret_hash,
+        ])[0]
+    }
+
+    fn compute_nullifier(self) -> Field {
+        // TODO(https://github.com/AztecProtocol/aztec-packages/issues/1386): should use
+        // `compute_note_hash_for_read_or_nullify` once public functions inject nonce!
+        let siloed_note_hash = compute_siloed_note_hash(TransparentNoteMethods, self);
+        // TODO(#1205) Should use a non-zero generator index.
+        pedersen([self.secret, siloed_note_hash])[0]
+    }
+
+    fn set_header(&mut self, header: NoteHeader) {
+        self.header = header;
+    }
+
+
+    // CUSTOM FUNCTIONS FOR THIS NOTE TYPE
+
+    fn compute_secret_hash(secret: Field) -> Field {
+        // TODO(#1205) This is probably not the right index to use
+        pedersen_with_separator([secret], GENERATOR_INDEX__L1_TO_L2_MESSAGE_SECRET)[0]
+    }
+
+    fn knows_secret(self, secret: Field) {
```

Comment:
Maybe add a comment to say that this is a custom method that's not enforced by the `NoteInterface`